doc/TEMPLATES.md

```diff
@@ -67,6 +67,10 @@ The following template languages are supported:
 * html
 * svg
 
+### Nested Data and Re-rendering
+
+Web components will only trigger a re-render by default **when their properties are replaced**. This means that when using non-primitive data structures as properties, mutating the contents of these structures will not result in the component re-rendering. This is by design. When you need to update a nested data structure on a web component, replace the property's value with a newly minted data structure.
+
```

Comment:
I think the language around “web components” is still a little too loose — this is more about template engines IMO. Rather than go back-and-forth, how about I just draft up a quick counter-proposal for you? Thoughts?

---

### A note on non-primitive data:
Because DOM manipulation is _slow_ — template engines do their best to avoid it (if a value _hasn’t changed_, DOM manipulation is skipped). To know if a non-primitive property has changed, a change-by-reference check is performed (versus a change-by-value check). Therefore _treat non-primitives as if they’re immutable_. I.e., do this: `element.property = { ...element.property, foo: 'bar' }`. And don’t do this: `element.property.foo = 'bar'`.

Reply:
@theengineear Thanks-- yes, some of the boundaries are still fuzzy to me since this project is my first use of web components. I was on the fence about writing something on the lines of 'treat non-primitives as though they're immutable', even though I was understanding this to be the implication, because that almost seemed like something that should be discussed in the design philosophy section, but I couldn't think of a way to fit it in there.

I'll think a bit more on whether there's a better way to communicate the expected design patterns this library is hoping to encourage. Although, it might be obvious to people who know they need it.

In any case, I made a couple of small grammatical changes to the suggested replacement, rebased and squashed. Should be ready for you, thanks!